riscv/csrs.cc

```diff
@@ -1240,3 +1240,110 @@ bool vxsat_csr_t::unlogged_write(const reg_t val) noexcept {
   dirty_vs_state;
   return masked_csr_t::unlogged_write(val);
 }
+
+// implement class stateen_csr_t
+stateen_csr_t::stateen_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask,
+                             const reg_t init):
+  masked_csr_t(proc, addr, mask, init) {
+}
+
+void stateen_csr_t::verify_permissions(insn_t insn, bool write) const {
+  if (!proc->extension_enabled(EXT_SMSTATEEN))
+    throw trap_illegal_instruction(insn.bits());
+  masked_csr_t::verify_permissions(insn, write);
+}
+
+// implement class stateenh_csr_t
+stateenh_csr_t::stateenh_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask,
+                               stateen_csr_t_p stateen):
+  csr_t(proc, addr),
+  stateen(stateen),
+  mask(mask) {
+}
+
+reg_t stateenh_csr_t::read() const noexcept {
+  return (stateen->read() >> 32) & mask;
+}
+
+bool stateenh_csr_t::unlogged_write(const reg_t val) noexcept {
+  return stateen->unlogged_write((stateen->written_value() & ~(mask << 32)) | ((val & mask) << 32));
+}
+
+void stateenh_csr_t::verify_permissions(insn_t insn, bool write) const {
+  stateen->verify_permissions(insn, write);
+}
+
+// implement class hstateen_csr_t
+hstateen_csr_t::hstateen_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask,
+                               const reg_t init, uint8_t index):
+  stateen_csr_t(proc, addr, mask, init),
+  index(index) {
+}
+
+reg_t hstateen_csr_t::read() const noexcept {
+  // For every bit in an mstateen CSR that is zero (whether read-only zero or set to zero),
+  // the same bit appears as read-only zero in the matching hstateen and sstateen CSRs
+  return stateen_csr_t::read() & state->mstateen[index]->read();
+}
+
+bool hstateen_csr_t::unlogged_write(const reg_t val) noexcept {
+  // For every bit in an mstateen CSR that is zero (whether read-only zero or set to zero),
+  // the same bit appears as read-only zero in the matching hstateen and sstateen CSRs
+  return stateen_csr_t:: unlogged_write(val & state->mstateen[index]->read());
+}
+
+void hstateen_csr_t::verify_permissions(insn_t insn, bool write) const {
+  if ((state->prv < PRV_M) && !(state->mstateen[index]->read() & MSTATEEN_HSTATEEN))
+    throw trap_illegal_instruction(insn.bits());
+
+  stateen_csr_t::verify_permissions(insn, write);
+}
+
+// implement class hstateenh_csr_t
+hstateenh_csr_t::hstateenh_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask,
+                                 stateen_csr_t_p stateen, uint8_t index):
+  stateenh_csr_t(proc, addr, mask, stateen),
+  index(index) {
+}
+
+void hstateenh_csr_t::verify_permissions(insn_t insn, bool write) const {
+  if ((state->prv < PRV_M) && !(state->mstateen[index]->read() & MSTATEEN_HSTATEEN))
+    throw trap_illegal_instruction(insn.bits());
```

Comment:
Won't this get checked when `stateenh_csr_t::verify_permissions()` calls `stateen->verify_permissions(insn, write);` which will call `hstateen_csr_t::verify_permissions()`?

In other words, it seems this entire method is unnecessary.

Reply:
Yeah. I'll delete this.